contracts/modules/cover/Cover.sol

```diff
@@ -186,343 +185,185 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard {
-    buyCoverParams.productId = cover.productId;
+      // existing cover
+      coverId = params.coverId;
+      require(ICoverNFT(coverNFT).isApprovedOrOwner(msg.sender, coverId), "Cover: Only owner or approved can edit");
 
-    Product memory product = _products[buyCoverParams.productId];
-    require(!product.isDeprecated, "Cover: Product is deprecated");
+      CoverData memory cover = _coverData[coverId];
+      require(params.coverAsset == cover.coverAsset, "Cover: unexpected coverAsset requested");
+      require(params.productId == cover.productId, "Cover: unexpected productId requested");
 
-    uint32 deprecatedCoverAssetsBitmap = pool().deprecatedCoverAssetsBitmap();
+      uint lastSegmentIndex = _coverSegments[coverId].length - 1;
+      CoverSegment memory lastSegment = coverSegments(coverId, lastSegmentIndex);
 
-    // Check that the payout asset is still supported (it may have been deprecated or disabled in the meantime)
-    require(
-      isAssetSupported(_getSupportedCoverAssets(deprecatedCoverAssetsBitmap, product.coverAssets), buyCoverParams.coverAsset),
-      "Cover: Payout asset is not supported"
-    );
-    require(
-      !_isCoverAssetDeprecated(deprecatedCoverAssetsBitmap, buyCoverParams.paymentAsset),
-      "Cover: payment asset deprecated"
-    );
-
-    uint refundInCoverAsset;
-
-    // if the last segment is not expired
-    if (lastCoverSegment.start + lastCoverSegment.period > uint32(block.timestamp)) {
-      refundInCoverAsset = deallocateCapacity(
-        coverId,
-        lastCoverSegmentIndex,
-        product,
-        cover,
-        lastCoverSegment
-      );
+      // if the last segment is not expired - make it end at the current block
+      if (lastSegment.start + lastSegment.period > block.timestamp) {
+        _coverSegments[coverId][lastSegmentIndex].period = (block.timestamp - lastSegment.start).toUint32();
+        // TODO: figure out how/where should we handle this
+        // tokenController().burnStakingPoolNXMRewards(deallocatedRewardsInNXM, allocation.poolId);
+      }
     }
 
-    (uint totalPremiumInNXM, uint totalCoveredAmountInCoverAsset) = _buyCover(buyCoverParams, coverId, poolAllocations);
-
-    handlePaymentAndRefund(buyCoverParams, totalPremiumInNXM, refundInCoverAsset);
+    {
+      // convert to NXM amount
+      uint nxmPriceInCoverAsset = pool().getTokenPrice(params.coverAsset);
+      uint segmentId = _coverSegments[coverId].length;
 
-    // Update total cover amount for asset if cover tracking is enabled
-    if (coverAmountTrackingEnabled) {
-       totalActiveCoverInAsset[cover.coverAsset] =
-        totalActiveCoverInAsset[cover.coverAsset] - lastCoverSegment.amount + totalCoveredAmountInCoverAsset;
-    }
+      uint totalCoverAmountInNXM;
+      uint premiumInNXM;
 
-    emit CoverEdited(coverId, cover.productId, lastCoverSegmentIndex + 1, msg.sender);
-  }
+      for (uint i = 0; i < allocationRequests.length; i++) {
```

Comment:
assuming the refund calculation is moved to the staking pool and you can only earn a refund within a particular pool, this loop looks cleaner as it is now